kafkakewl-common/src/main/scala/com/mwam/kafkakewl/common/kafka/KafkaConsumerUtils.scala

```diff
@@ -53,130 +72,75 @@ object KafkaConsumerExtensions {
-    noOfConsumedMessages: Long
-) {
-  @targetName("add")
-  def +(other: CompactedConsumeResult[Key, Value]): CompactedConsumeResult[Key, Value] = CompactedConsumeResult(
-    lastValues ++ other.lastValues,
-    nextOffsets ++ other.nextOffsets,
-    noOfConsumedMessages + other.noOfConsumedMessages
-  )
-}
-
-object KafkaConsumerUtils {
-  import KafkaConsumerExtensions.*
-  import KafkaClientConfigExtensions.*
-
-  def kafkaConsumer[Key, Value](
-      kafkaClientConfig: KafkaClientConfig,
-      keyDeserializer: Deserializer[Key],
-      valueDeserializer: Deserializer[Value]
-  ): RIO[Scope, Consumer[Key, Value]] = {
-    ZIO.fromAutoCloseable(
-      ZIO.attempt(
-        new KafkaConsumer[Key, Value](
-          kafkaClientConfig.toProperties,
-          keyDeserializer,
-          valueDeserializer
-        )
-      )
-    )
-  }
-
-  def kafkaConsumerStringStringZIO(kafkaClientConfig: KafkaClientConfig): RIO[Scope, Consumer[String, String]] =
-    kafkaConsumer(kafkaClientConfig, new StringDeserializer(), new StringDeserializer())
-
-  def kafkaConsumerBytesBytesZIO(kafkaClientConfig: KafkaClientConfig): RIO[Scope, Consumer[Bytes, Bytes]] =
-    kafkaConsumer(kafkaClientConfig, new BytesDeserializer(), new BytesDeserializer())
-
   def consumeUntilEnd[Key, Value](
-      consumer: Consumer[Key, Value],
+      consumer: Consumer,
       topicPartitions: List[TopicPartition],
+      keySerializer: Deserializer[Any, Key],
+      valueSerializer: Deserializer[Any, Value],
       pollTimeout: Duration = 100.millis
-  )(
-      handleConsumerRecords: ConsumerRecords[Key, Value] => Unit
-  ): Map[TopicPartition, Long] = {
-
-    val pollTimeoutJava = java.time.Duration.ofNanos(pollTimeout.toNanos)
-    val topicPartitionsJava = topicPartitions.asJava
-
-    consumer.assign(topicPartitionsJava)
-    consumer.seekToBeginning(topicPartitionsJava)
-    val endOffsets = consumer.endOffsets(topicPartitions.toSet)
-
-    var shouldConsume = true
-    val nextTopicPartitionOffsets = mutable.Map.empty[TopicPartition, Long] ++ consumer.positions(topicPartitions)
-    while (shouldConsume) {
-      val consumerRecords = consumer.poll(pollTimeoutJava)
-      val nextOffsets = consumer.positions(topicPartitions)
-      shouldConsume = endOffsets.exists { case (tp, endOffset) => nextOffsets.get(tp).exists(nextOffset => nextOffset < endOffset) }
-      handleConsumerRecords(consumerRecords)
-      consumerRecords.forEach { cr =>
-        nextTopicPartitionOffsets.update(new TopicPartition(cr.topic, cr.partition), cr.offset + 1L)
-      }
-    }
+  ): Task[(Map[TopicPartition, Long], List[ConsumerRecord[Key, Value]])] = {
+    val subscription = Manual(topicPartitions.toSet)
+    for {
+      endOffsets <- consumer.endOffsets(topicPartitions.toSet)
+      partitionMapAndRecords <- consumer
+        .plainStream(subscription, keySerializer, valueSerializer)
+        .aggregateAsyncWithin(ZSink.collectAll, Schedule.fixed(pollTimeout))
```

Comment:
I've changed it and merged it with a ZStream that produces an empty chunk

Reply:
This worries me a little bit. I thought that `aggregateAsyncWithin()` will emit an empty chunk after every `pollTimeout` interval. If that's not the case, I'm not sure if merging it with one empty chunk always solves it.

The empty chunk is emitted very early one, possibly before the first polling of the consumer. Maybe at that point the position isn't up to date yet. In the unlikely extreme cases we may have lot of rolled back messages in the topic that multiple poll can skip and update the position to the end-offset. In this case would this work?

Conceptually we periodically need to check the position whether it reached the end-offsets. Possibly a completely separate repeated ZIO could do that and it would signal the completion to the main stream?